core/rawdb/accessors_chain.go

```diff
@@ -823,6 +823,80 @@ func WriteAncientBlocks(db ethdb.AncientWriter, blocks []*types.Block, receipts
 	})
 }
 
+// WriteAncientBlocksAfterCancun writes entire block data into ancient store and returns the total written size.
+func WriteAncientBlocksAfterCancun(db ethdb.Database, config *params.ChainConfig, blocks []*types.Block, receipts []types.Receipts, td *big.Int) (int64, error) {
+	if len(blocks) == 0 {
+		return 0, nil
+	}
+
+	cancunIndex := -1
+	for i, block := range blocks {
+		if config.IsCancun(block.Number(), block.Time()) {
+			cancunIndex = i
+			break
+		}
+	}
+	if cancunIndex < 0 {
+		return WriteAncientBlocks(db, blocks, receipts, td)
+	}
+
+	// if there has pre-cancun and post-cancun blocks, write them separately
+	var (
+		writeSize  int64
+		tdSum      = new(big.Int).Set(td)
+		stReceipts []*types.ReceiptForStorage
+	)
+
+	// write pre-cancun blocks
+	if cancunIndex > 0 {
+		preBlocks := blocks[:cancunIndex]
+		preReceipts := receipts[:cancunIndex]
+		size, err := db.ModifyAncients(func(op ethdb.AncientWriteOp) error {
+			for i, block := range preBlocks {
```

Comment:
854~870 call `WriteAncientBlocks` instead?

Reply:
Oh, here is for sharing the TD.